test_unstructured/file_utils/test_filetype.py

```diff
@@ -548,3 +550,260 @@ def magic_from_buffer_(request: FixtureRequest):
+            ctx = _FileTypeDetectionContext(file_path=example_doc_path("simple.doc"))
+            assert ctx.mime_type == "application/msword"
+
+    def but_it_warns_to_install_libmagic_when_the_filetype_lib_cannot_detect_the_MIME_type(
+        self, caplog: LogCaptureFixture
+    ):
+        with patch("unstructured.file_utils.filetype.LIBMAGIC_AVAILABLE", False):
+            ctx = _FileTypeDetectionContext(file_path=example_doc_path("norwich-city.txt"))
+            assert ctx.mime_type is None
+            assert "WARNING" in caplog.text
+            assert "libmagic is unavailable" in caplog.text
+            assert "consider installing libmagic" in caplog.text
+
+    def it_provides_the_MIME_type_detected_by_libmagic_from_a_file_like_object(self):
+        with open(example_doc_path("norwich-city.txt"), "rb") as f:
+            ctx = _FileTypeDetectionContext(file=f)
+            assert ctx.mime_type == "text/plain"
+
+    def and_it_provides_the_MIME_type_from_file_using_filetype_lib_when_magic_is_unavailable(self):
+        with patch("unstructured.file_utils.filetype.LIBMAGIC_AVAILABLE", False):
+            file_path = example_doc_path("simple.doc")
+            with open(file_path, "rb") as f:
+                ctx = _FileTypeDetectionContext(file=f)
+                assert ctx.mime_type == "application/msword"
+
+    # -- .open() ------------------------------------------------
+
+    def it_provides_transparent_access_to_the_source_file_when_it_is_a_file_like_object(self):
+        with open(example_doc_path("norwich-city.txt"), "rb") as f:
+            ctx = _FileTypeDetectionContext(file=f)
+            with ctx.open() as file:
+                assert file is f
+                assert file.read(38) == b"Iwan Roberts\nRoberts celebrating after"
+
+    def it_provides_transparent_access_to_the_source_file_when_it_is_a_file_path(self):
+        ctx = _FileTypeDetectionContext(file_path=example_doc_path("norwich-city.txt"))
+        with ctx.open() as file:
+            assert file.read(38) == b"Iwan Roberts\nRoberts celebrating after"
+
+    # -- .text_head ---------------------------------------------
+
+    def it_grabs_the_first_4k_chars_from_file_path_for_textual_type_differentiation(self):
+        ctx = _FileTypeDetectionContext(file_path=example_doc_path("norwich-city.txt"))
+
+        text_head = ctx.text_head
+
+        assert isinstance(text_head, str)
+        assert len(text_head) == 4096
+        assert text_head.startswith("Iwan Roberts\nRoberts celebrating after")
+
+    def and_it_uses_character_detection_to_correct_a_wrong_encoding_arg_for_file_path(self):
+        ctx = _FileTypeDetectionContext(
+            file_path=example_doc_path("norwich-city.txt"), encoding="utf_32_be"
+        )
+
+        text_head = ctx.text_head
+
+        assert isinstance(text_head, str)
+        assert len(text_head) == 4096
+        assert text_head.startswith("Iwan Roberts\nRoberts celebrating after")
```

Comment:
Should there be an assert using `ctx.encoding` here?

Reply:
No, I'd say not; the behavior this test is checking is more like "it can still read the text file, at least in this case, when the encoding asserted by the caller is wrong". We could have instead mocked `detect_file_encoding()` but I thought it was better to retain that integration in the test while we exercise the `UnicodeDecodeError` code-path. 

How the code deeper down gets that done is not under test here. I do think it's worth testing over there but I assume we have some tests for that already and anyway that's outside the scope of this work.